phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -953,20 +973,39 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
                     allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, SortOrder.getDefault()));
-                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault()));
+                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault(), null));
                 }
                 // First columns are the indexed ones
-                for (Pair<ColumnName, SortOrder> pair : indexedPkColumns) {
-                    ColumnName colName = pair.getFirst();
-                    PColumn col = resolver.resolveColumn(null, colName.getFamilyName(), colName.getColumnName()).getColumn();
-                    unusedPkColumns.remove(col);
-                    // Ignore view constants for updatable views as we don't need these in the index
-                    if (col.getViewConstant() == null) {
-                        PDataType dataType = IndexUtil.getIndexColumnDataType(col);
-                        colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
-                        allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, pair.getSecond()));
-                        columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), col.isNullable(), col.getMaxLength(), col.getScale(), false, SortOrder.getDefault()));
-                    }
+                for (Pair<ParseNode, SortOrder> pair : indexParseNodeAndSortOrderList) {
+                	ParseNode parseNode = pair.getFirst();
+                	ColumnName colName = null;
+                	Integer maxLength = null;
+                	Integer scale = null;
+                	// if it is a column 
+                	if (parseNode instanceof ColumnParseNode) {
+                		ColumnParseNode colParseNode = (ColumnParseNode)parseNode;
+	                    PColumn col = resolver.resolveColumn(null, colParseNode.getTableName(), colParseNode.getName()).getColumn();
+	                    unusedPkColumns.remove(col);
+	                    // Ignore view constants for updatable views as we don't need these in the index
+	                    if (col.getViewConstant() != null) 
+	                    	continue;
+	                    colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
+	                    maxLength = col.getMaxLength();
+	                    scale = col.getScale();
+                	}
+                	// compile the parseNode to get an expression
+			        PhoenixStatement phoenixStatment = new PhoenixStatement(connection);
+					final StatementContext context = new StatementContext(phoenixStatment, resolver);
+			        ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+			        Expression expression = parseNode.accept(expressionCompiler);	
```

Comment:
I think here you'll want to check if expressionCompiler.isAggregate() and throw here if that's the case. It wouldn't be valid to use aggregate functions in a functional index.